Haiku: Process/thread management functions

trungnt2910 opened the pull request:
Add support for process/thread management functions in `System.Diagnostics.Process` for Haiku.

This is required to build managed runtime libraries for Haiku as well as running a simple "Hello, World!" application.

Part of #55803.

trungnt2910 committed:
Haiku: Process/thread management functions

Add support for process/thread management functions in
`System.Diagnostics.Process` for Haiku.

This is required to build managed runtime libraries for Haiku as well as
running a simple "Hello, World!" application.

Co-authored-by: Jessica Hamilton <jessica.l.hamilton@gmail.com>

trungnt2910 commented:
C/c @am11 

I just added the relevant parts of the common files to both, then we can rebase stuff later.

trungnt2910 commented on src/libraries/Common/src/Interop/Haiku/Interop.OS.cs:
```diff
@@ -0,0 +1,257 @@
+        /// Gets information about a team's usage.
+        /// </summary>
+        /// <param name="team">The team ID.</param>
+        /// <param name="who">The thread ID</param>
```
Lol that's probably me unconsciously accepting Copilot's suggestions in 2023 :trollface:

trungnt2910 commented on src/libraries/Common/src/Interop/Haiku/Interop.Image.cs:
```diff
@@ -0,0 +1,58 @@
+            public int sequence;
+            public int init_order;
+            public delegate* unmanaged<void> init_routine;
+            public delegate* unmanaged<void> term_routine;
```
Maybe an opaque `void*` instead, since the other members are `void*`?

Those two correspond to https://github.com/haiku/haiku/blob/d39ce117e02d3d8aac966b1a46bd955a2b59c069/headers/os/kernel/image.h#L27-L28 in the Haiku source.

```c
	void		(*init_routine)();
	void		(*term_routine)();
```

huoyaoyuan replied:
Semantic-wise `void*` is incorrect because function pointers can't be dereferenced. In C function pointers can't be casted to `void*` either. `IntPtr` is correct for opaque pointer-size field. Unmanaged function pointer is precise but unnecessary.

jeffhandley replied:
@trungnt2910 It looks like this thread is still unresolved with the request to use `IntPtr`

trungnt2910 replied:
I removed the relevant members entirely after following the approach in #121880.

These two problematic members are not referenced from the managed code.

trungnt2910 committed:
Haiku: Declare as supported platform

Declare Haiku as a supported platform to MSBuild for all
`System.Diagnostics.Process` builds.

This prevents `CA1418` when using the `SupportedOSPlatform` attribute
with `haiku`.

trungnt2910 committed:
Haiku: Align attributes with reference API defs

jeffhandley commented:
@trungnt2910 I pushed https://github.com/dotnet/runtime/commit/0ee9ed30d7803b1867892acea17e4aaefb00524b to my fork; it merges main into your branch and resolves the conflict. Feel free to cherry-pick that commit if it's helpful.

trungnt2910 commented on src/libraries/Common/src/Interop/Haiku/Interop.OS.cs:
```diff
@@ -4,25 +4,164 @@
+        /// </summary>
+        /// <param name="thread">The thread ID.</param>
+        /// <param name="newPriority">The new priority.</param>
+        /// <returns>The previous priority if successful or an error code on failure.</returns>
```
Done. Fixed this not by applying the docs but fix the underlying behavior of the native code and its consumers.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Haiku.cs:
```diff
@@ -0,0 +1,224 @@
+
+            return processes.ToArray();
+        }
+
```
Done. Refactored to match with the latest .NET internal `Process` APIs.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Haiku.cs:
```diff
@@ -0,0 +1,205 @@
+            }
+
+            return processes.ToArray();
+        }
```
Done. Refactored to match with the latest .NET internal `Process` APIs.

trungnt2910 commented on src/native/libs/configure.cmake:
```diff
@@ -1004,6 +1004,10 @@ check_include_files(
     IOKit/serial/ioss.h
     HAVE_IOSS_H)
 
+check_include_files(
+    image.h
+    HAVE_IMAGE_H)
```
Disagree, as discussed in #121880.

trungnt2910 commented on src/native/libs/System.Native/pal_getosinfo.c:
```diff
@@ -12,6 +12,37 @@
 #include <OS.h>
 #endif
 
+#if HAVE_IMAGE_H
+#include <image.h>
+#endif
```
Disagree, as discussed in #121880.

trungnt2910 commented on src/native/libs/configure.cmake:
```diff
@@ -1004,6 +1004,10 @@ check_include_files(
     IOKit/serial/ioss.h
     HAVE_IOSS_H)
 
+check_include_files(
+    image.h
+    HAVE_IMAGE_H)
+
```
Disagree.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Haiku.cs:
```diff
@@ -0,0 +1,196 @@
+            return processInfo?.ProcessName;
+        }
+
+        internal static ProcessInfo? CreateProcessInfo(int pid, string? processNameFilter = null)
+        {
+            // Negative PIDs aren't valid
+            ArgumentOutOfRangeException.ThrowIfNegative(pid);
+
+            Interop.OS.TeamInfo info;
+            int status = Interop.OS.GetTeamInfo(pid, out info);
+
+            if (status != 0)
```
Haven't seen any build errors so far.

trungnt2910 commented on src/native/libs/System.Native/pal_getosinfo.c:
```diff
@@ -40,3 +71,231 @@ int32_t SystemNative_GetNextAreaInfo(int32_t team, intptr_t* cookie, AreaInfo* a
+#endif
+}
+
+int32_t SystemNative_GetNextImageInfo(int32_t team, int32_t* cookie, ImageInfo* info)
+{
+#if HAVE_IMAGE_H
+    if (cookie == NULL || info == NULL)
+    {
+        return EINVAL;
```
Done, added the guard.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Haiku.cs:
```diff
@@ -0,0 +1,196 @@
+        /// <summary>Gets process infos for each process on the local machine.</summary>
+        /// <param name="builder">The builder to add found process infos to.</param>
+        /// <param name="processNameFilter">Optional process name to use as an inclusion filter.</param>
+        public static void GetProcessInfos(ref ArrayBuilder<ProcessInfo> builder, string? processNameFilter)
+        {
+            int cookie = 0;
+            Interop.OS.TeamInfo info;
+
+            while ((Interop.OS.GetNextTeamInfo(ref cookie, out info)) == 0)
+            {
+                ProcessInfo? pi = GetProcessInfoFromTeamInfo(ref info, processNameFilter);
```
Haven't seen any build errors so far.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Haiku.cs:
```diff
@@ -0,0 +1,197 @@
+        }
+
+        /// <summary>Gets the time the associated process was started.</summary>
+        internal DateTime StartTimeCore
+        {
+            get
+            {
+                EnsureState(State.HaveNonExitedId);
+
+                Interop.OS.TeamInfo info;
+                int status = Interop.OS.GetTeamInfo(_processId, out info);
+
```
Haven't seen any build errors so far.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Haiku.cs:
```diff
@@ -0,0 +1,197 @@
+                if (status != 0)
+                {
+                    throw new Win32Exception(status);
+                }
+
+                return BootTime.AddTicks(info.start_time * BigTimeToTicks);
+            }
```
Fixed in the latest commit.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Haiku.cs:
```diff
@@ -0,0 +1,197 @@
+
+            while ((Interop.Image.GetNextImageInfo(processId, ref cookie, out info)) == 0)
+            {
+                string modulePath = GetString(info.name, Interop.Image.MAXPATHLEN);
```
Haven't seen any errors so far.

trungnt2910 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Haiku.cs:
```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+
```
?? Are you hallucinating? There is no `System.Collections.Generic;` let me have a break won't you?

trungnt2910 commented on src/native/libs/System.Native/pal_getosinfo.c:
```diff
@@ -40,3 +71,231 @@ int32_t SystemNative_GetNextAreaInfo(int32_t team, intptr_t* cookie, AreaInfo* a
+
+    memset(info, 0, sizeof(*info));
+    info->type = nativeInfo.type;
+    memcpy(info->name, nativeInfo.name, MAXPATHLEN);
```
Done, used `sizeof(info->name)` and used `SafeStringCopy` instead. I definitely did prompt you to use the safe string copy function locally but you ignored anyway. 😠

trungnt2910 commented on src/native/libs/System.Native/pal_getosinfo.c:
```diff
@@ -12,6 +12,37 @@
+    info->parent = nativeInfo->parent;
+    memcpy(info->name, nativeInfo->name, sizeof(info->name));
+    info->start_time = nativeInfo->start_time;
+}
+
+static void CopyThreadInfo(HaikuThreadInfo* info, const thread_info* nativeInfo)
+{
+    info->thread = nativeInfo->thread;
+    info->team = nativeInfo->team;
+    info->state = nativeInfo->state;
+    info->priority = nativeInfo->priority;
+    info->user_time = nativeInfo->user_time;
+    info->kernel_time = nativeInfo->kernel_time;
+}
```
Done.

trungnt2910 commented:
Local builds are succeeding in both Release and Debug mode (with all my other open PRs on native components applied).

jkotas commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Haiku.cs:
```diff
@@ -0,0 +1,197 @@
+                length++;
+            }
+
+            return new string((sbyte*)ptr, 0, length);
```
```suggestion
            int length = new ReadOnlySpan<byte>(ptr, maxLength).IndexOf(0);
            return new string((sbyte*)ptr, 0, (length < 0) ? maxLength : length);
```

am11 commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Haiku.cs:
```diff
@@ -0,0 +1,192 @@
+
+        private static unsafe string GetString(byte* ptr, int maxLength)
+        {
+            int length = new ReadOnlySpan<byte>(ptr, maxLength).IndexOf(0);
```
```suggestion
            int length = new ReadOnlySpan<byte>(ptr, maxLength).IndexOf((byte)0);
```
to fix build break.

am11 commented:
LGTM!
cc @adamsitnik, PTAL.

trungnt2910 committed:
Haiku: Add missing enum-to-int casts

Otherwise `clang`-based builds will complain.

`clang` treats the `enum`s as unsigned `int`, triggering a sign conversion warning.
